File: source/com/barrybecker4/simulation/fluid/rendering/EnvironmentRenderer.java
/** Copyright by Barry G. Becker, 2000-2011. Licensed under MIT License: http://www.opensource.org/licenses/MIT  */
package com.barrybecker4.simulation.fluid.rendering;

import com.barrybecker4.simulation.common.rendering.ModelImage;
import com.barrybecker4.simulation.fluid.model.Grid;
import com.barrybecker4.ui.util.ColorMap;

import java.awt.Color;
import java.awt.Graphics2D;
import java.util.ArrayList;
import java.util.List;

/**
 *  Renders a specified liquid environment.
 *
 *  @author Barry Becker
 */
public final class EnvironmentRenderer {

    // rendering attributes
    private static final Color GRID_COLOR = new Color( 30, 30, 30, 10 );
    private static final Color VECTOR_COLOR = new Color( 200, 60, 30, 50 );


    private static final double  VECTOR_SCALE = 40.0;
    private static final int OFFSET = 10;

    private static final PressureColorMap PRESSURE_COLOR_MAP = new PressureColorMap();

    private Grid grid;
    private ModelImage modelImage;
    private RenderingOptions options;


    /** Constructor */
    public EnvironmentRenderer(Grid grid, RenderingOptions options) {

        this.grid = grid;
        this.options = options;
        modelImage = new ModelImage(grid, PRESSURE_COLOR_MAP, (int)options.getScale());
    }

    public ColorMap getColorMap() {
        return PRESSURE_COLOR_MAP;
    }

    public RenderingOptions getOptions() {
        return options;
    }

    /**
     * Render the Environment on the screen.
     */
    public void render(Graphics2D g) {

        // draw the cells colored by ---pressure--- val
        if (options.getShowPressures()) {
            concurrentRenderPressures(g);
        }

        // outer boundary
        double scale = options.getScale();
        g.drawRect( OFFSET, OFFSET, (int) (grid.getWidth() * scale), (int) (grid.getHeight() * scale) );

        // draw the ---velocity--- field (and status)
        if (options.getShowVelocities()) {
            drawVectors(g);
        }

        if (options.getShowGrid())  {
            drawGrid(g);
        }
    }

    /**
     * If the render options say to use parallelism, then we will render the pressures concurrently.
     */
    private void concurrentRenderPressures(Graphics2D g2) {

        int height = grid.getHeight();
        modelImage.setUseLinearInterpolation(options.getUseLinearInterpolation());

        int numProcs = options.getParallelizer().getNumThreads();
        List<Runnable> workers = new ArrayList<Runnable>(numProcs);
        int range = (height / numProcs);
        for (int i = 0; i < (numProcs - 1); i++) {
            int offset = i * range;
            workers.add(new RenderWorker(modelImage, offset, offset + range));
        }
        // leftover in the last strip, or all of it if only one processor.
        int currentRow = range * (numProcs - 1);
        workers.add(new RenderWorker(modelImage, currentRow, height));

        // blocks until all Callables are done running.
        options.getParallelizer().invokeAllRunnables(workers);

        g2.drawImage(modelImage.getImage(), OFFSET, OFFSET, null);
    }

    private void drawGrid(Graphics2D g)    {
        g.setColor( GRID_COLOR );
        double scale = options.getScale();

        int rightEdgePos = (int) (scale * grid.getWidth());
        int bottomEdgePos = (int) (scale * grid.getHeight());

        for (int j = 0; j < grid.getHeight(); j++ )   //  -----
        {
            int ypos = (int) (j * scale);
            g.drawLine( OFFSET, ypos + OFFSET, rightEdgePos + OFFSET, ypos + OFFSET );
        }
        for (int i = 0; i < grid.getWidth(); i++ )    //  ||||
        {
            int xpos = (int) (i * scale);
            g.drawLine( xpos + OFFSET, OFFSET, xpos + OFFSET, bottomEdgePos + OFFSET );
        }
    }


    private void drawVectors(Graphics2D g) {
        g.setColor( VECTOR_COLOR );
        double scale = options.getScale();

        for ( int j = 0; j < grid.getHeight(); j++ ) {
            for ( int i = 0; i < grid.getWidth(); i++ ) {
                double u = grid.getU(i, j);
                double v = grid.getV(i, j);
                int x = (int) (scale * i) + OFFSET;
                int y = (int) (scale * j) + OFFSET;

                g.drawLine( x, y,
                        (int) (scale * i + VECTOR_SCALE  * u) + OFFSET, (int) (scale * j + VECTOR_SCALE * v) + OFFSET );

            }
        }
    }
}

File: scala-source/com/barrybecker4/simulation/fluid/rendering/EnvironmentRenderer.scala
// Copyright by Barry G. Becker, 2016-2017. Licensed under MIT License: http://www.opensource.org/licenses/MIT
package com.barrybecker4.simulation.fluid.rendering

import com.barrybecker4.simulation.common.rendering.RectangularModelImage
import com.barrybecker4.simulation.fluid.model.FluidEnvironment
import java.awt.Color
import java.awt.Graphics2D
import EnvironmentRenderer._
import scala.collection.parallel.CollectionConverters._

/**
  * Renders a specified fluid environment.
  * @author Barry Becker
  */
object EnvironmentRenderer {
  val ContentOffsetPixels: Int = 10
  private val GRID_COLOR = new Color(30, 30, 30, 10)
  private val VECTOR_COLOR = new Color(200, 60, 30, 50)
  private val VECTOR_SCALE = 40.0
  private val PRESSURE_COLOR_MAP = new PressureColorMap
}

final class EnvironmentRenderer(var env: FluidEnvironment, var options: RenderingOptions) {

  private val modelImage = new RectangularModelImage(env, PRESSURE_COLOR_MAP, options.getScale.toInt)

  def getColorMap: PressureColorMap = PRESSURE_COLOR_MAP
  def getOptions: RenderingOptions = options

  /** Render the Environment on the screen. */
  def render(g: Graphics2D): Unit = {
    // draw the cells colored by ---pressure--- val
    if (options.getShowPressures) concurrentRenderPressures(g)
    // outer boundary
    val scale = options.getScale
    g.drawRect(ContentOffsetPixels,
      ContentOffsetPixels, (env.getWidth * scale).toInt, (env.getHeight * scale).toInt)

    // draw the ---velocity--- field (and status)
    if (options.getShowVelocities) drawVectors(g)
    if (options.getShowGrid) drawGrid(g)
  }

  /** If the render options say to use parallelism, then we will render the pressures concurrently. */
  private def concurrentRenderPressures(g2: Graphics2D): Boolean = {
    val height = env.getHeight
    modelImage.setUseLinearInterpolation(options.getUseLinearInterpolation)
    val numProcs = Runtime.getRuntime.availableProcessors()

    val workers = Array.ofDim[Runnable](numProcs)
    val range = height / numProcs
    for (i <- 0 until numProcs - 1) {
      val offset = i * range
      workers(i) = new RenderWorker(modelImage, offset, offset + range)
    }
    // leftover in the last strip, or all of it if only one processor.
    val currentRow = range * (numProcs - 1)
    workers(numProcs - 1)  = new RenderWorker(modelImage, currentRow, height)

    // blocks until all Callables are done running.
    if (options.isParallelized) workers.par.foreach(_.run())
    else workers.foreach(_.run())

    g2.drawImage(modelImage.getImage, ContentOffsetPixels, ContentOffsetPixels, null)
  }

  private def drawGrid(g: Graphics2D): Unit = {
    g.setColor(GRID_COLOR)
    val scale = options.getScale
    val rightEdgePos = (scale * env.getWidth).toInt
    val bottomEdgePos = (scale * env.getHeight).toInt
    for (j <- 0 until env.getHeight) { //  -----
      val ypos = (j * scale).toInt
      g.drawLine(ContentOffsetPixels, ypos + ContentOffsetPixels, rightEdgePos + ContentOffsetPixels, ypos + ContentOffsetPixels)
    }
    for (i <- 0 until env.getWidth) { //  ||||
      val xpos = (i * scale).toInt
      g.drawLine(xpos + ContentOffsetPixels,
        ContentOffsetPixels, xpos + ContentOffsetPixels, bottomEdgePos + ContentOffsetPixels)
    }
  }

  private def drawVectors(g: Graphics2D): Unit = {
    g.setColor(VECTOR_COLOR)
    val scale = options.getScale
    for (j <- 0 until env.getHeight) {
      for (i <- 0 until env.getWidth) {
        val u = env.getU(i, j)
        val v = env.getV(i, j)
        val x = (scale * i).toInt + ContentOffsetPixels
        val y = (scale * j).toInt + ContentOffsetPixels
        g.drawLine(x, y,
          (scale * i + VECTOR_SCALE * u).toInt + ContentOffsetPixels,
          (scale * j + VECTOR_SCALE * v).toInt + ContentOffsetPixels)
      }
    }
  }
}
